Plugins/NativeCamera/Source/Apple/NativeCameraImpl.mm

```diff
@@ -185,198 +173,204 @@ fragment float4 fragmentShader(RasterizerData in [[stage_in]],
-                    std::make_exception_ptr(std::runtime_error{"ConstraintError: Unable to match constraints to a supported camera configuration."})));
-                return;
-            }
+        UNUSED(maxWidth);
+        UNUSED(maxHeight);
+        UNUSED(frontCamera);
+        AVCaptureDevice* captureDevice{[AVCaptureDevice defaultDeviceWithMediaType:AVMediaTypeVideo]};
+        AVCaptureDeviceInput *input{[AVCaptureDeviceInput deviceInputWithDevice:captureDevice error:&error]};
+        CMVideoFormatDescriptionRef videoFormatRef{static_cast<CMVideoFormatDescriptionRef>(captureDevice.activeFormat.formatDescription)};
+        CMVideoDimensions dimensions{CMVideoFormatDescriptionGetDimensions(videoFormatRef)};
+        uint32_t devicePixelFormat{static_cast<uint32_t>(CMFormatDescriptionGetMediaSubType(videoFormatRef))};
+        if (!isPixelFormatSupported(devicePixelFormat))
+        {
+            return arcana::task_from_error<CameraDimensions>(std::make_exception_ptr(std::runtime_error{"ConstraintError: Unable to match constraints to a supported camera configuration."}));
+        }
 #endif
 
-            Camera::Impl::CameraDimensions cameraDimensions{static_cast<uint32_t>(dimensions.width), static_cast<uint32_t>(dimensions.height)};
-            
-            // For portrait orientations swap the height and width of the video format dimensions.
-            if (m_implData->cameraTextureDelegate->videoOrientation == AVCaptureVideoOrientationPortrait
-                ||  m_implData->cameraTextureDelegate->videoOrientation == AVCaptureVideoOrientationPortraitUpsideDown)
-            {
-                std::swap(cameraDimensions.width, cameraDimensions.height);
-            }
-            
-            // Check for failed initialisation.
-            if (!input)
-            {
-                taskCompletionSource.complete(arcana::make_unexpected(std::make_exception_ptr(std::runtime_error{"Error Getting Camera Input"})));
-                return;
+        Camera::Impl::CameraDimensions cameraDimensions{static_cast<uint32_t>(dimensions.width), static_cast<uint32_t>(dimensions.height)};
+
+        // For portrait orientations swap the height and width of the video format dimensions.
+        if (m_implData->cameraTextureDelegate->videoOrientation == AVCaptureVideoOrientationPortrait
+            ||  m_implData->cameraTextureDelegate->videoOrientation == AVCaptureVideoOrientationPortraitUpsideDown)
+        {
+            std::swap(cameraDimensions.width, cameraDimensions.height);
+        }
+        
+        // Check for failed initialisation.
+        if (!input)
+        {
+            return arcana::task_from_error<CameraDimensions>(std::make_exception_ptr(std::runtime_error{"Error Getting Camera Input"}));
+        }
+
+        // Kick off camera session on a background thread.
+        return arcana::make_task(m_cameraSessionDispatcher, arcana::cancellation::none(), [this, input, devicePixelFormat, cameraDimensions]() mutable {
```

Comment:
Is it possible for this class to be destroyed while an async operation is still in progress? Instead of capturing `this`, do we need to capture a shared_ptr (since this class already inherits from `enable_shared_from_this`)?

Reply:
I don't think it's possible in the current integration in BRN, but it could be possible for other integrations, so this is a good suggestion for safety, will look into it.